mne/preprocessing/ica.py

```diff
@@ -1151,6 +1153,101 @@ def find_bads_ecg(self, inst, ch_name=None, threshold=None, start=None,
+            The value above which a feature is classified as outlier.
+        start : int | float | None
+            First sample to include. If float, data will be interpreted as
+            time in seconds. If None, data will be used from the first sample.
+        stop : int | float | None
+            Last sample to not include. If float, data will be interpreted as
+            time in seconds. If None, data will be used to the last sample.
+        l_freq : float
+            Low pass frequency.
+        h_freq : float
+            High pass frequency.
+        reject_by_annotation : bool
+            If True, data annotated as bad will be omitted. Defaults to True.
+
+            .. versionadded:: 0.14.0
+
+        verbose : bool, str, int, or None
+            If not None, override default verbose level (see
+            :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>`
+            for more). Defaults to self.verbose.
+
+        Returns
+        -------
+        ref_idx : list of int
+            The indices of MEG reference related components, sorted by score.
+        scores : np.ndarray of float, shape (``n_components_``) | list of array
+            The correlation scores.
+
+        See Also
+        --------
+        find_bads_ecg, find_bads_eog
+        """
+        if verbose is None:
+            verbose = self.verbose
+
+        if not picks:
+            inds = pick_channels_regexp(inst.ch_names,"REF_ICA*")
+        else:
+            inds = picks
+        scores, ref_idx = [], []
+        ref_chs = [inst.ch_names[k] for k in inds]
+
+        # some magic we need inevitably ...
+        # get targets before equalizing
+        targets = [self._check_target(k, inst, start, stop,reject_by_annotation) for k in ref_chs]
+
+        for ii, (ref_ch, target) in enumerate(zip(ref_chs, targets)):
+            scores += [self.score_sources(
+                inst, target=target, score_func='pearsonr', start=start,
+                stop=stop, l_freq=l_freq, h_freq=h_freq, verbose=verbose,
+                reject_by_annotation=reject_by_annotation)]
+            # pick last scores
+            this_idx = find_outliers(scores[-1], threshold=threshold)
+            ref_idx += [this_idx]
+            self.labels_[('ref/%i/' % ii) + ref_ch] = list(this_idx)
+
+        # remove duplicates but keep order by score, even across multiple
+        # EOG channels
+        scores_ = np.concatenate([scores[ii][inds]
+                                  for ii, inds in enumerate(ref_idx)])
+        ref_idx_ = np.concatenate(ref_idx)[np.abs(scores_).argsort()[::-1]]
+
+        ref_idx_unique = list(np.unique(ref_idx_))
+        ref_idx = []
+        for i in ref_idx_:
+            if i in ref_idx_unique:
+                ref_idx.append(i)
+                ref_idx_unique.remove(i)
+        if len(scores) == 1:
+            scores = scores[0]
+        self.labels_['ref'] = list(ref_idx)
+
+        return self.labels_['ref'], scores
```

Comment:
How much of this code is shared with `find_bads_eog` / `find_bads_ecg`? If it's 95%+, then we should refactor so that this and those functions each just call a `return self._find_bads_type(...)` to reduce code duplication.

Reply:
@jshanna100 you should then refactor to have a private function eg _find_bads_ch_type that takes the channel type you want to correlate with and make find_bads_ecg and find_bads_ref_meg use this private function. This would avoid the code duplication.

also you can run:

`make pep`

from the mne-python folder. It will show you the code formatting errors that prevent travisci to be green.

thanks a lot for your efforts.

Reply:
Ok, this is done, I think. It works on some of my own data with reference channels. I've run some quick tests on the MNE sample data, and as far as I can tell, the ECG and EOG functions also seem to work as they should.